packages/sdk-js/src/handlers/auth.ts

```diff
@@ -42,45 +48,74 @@ export class AuthHandler {
 	}
 
 	async login(credentials: LoginCredentials) {
+		clearTimeout(this.autoRefreshTimeout);
 		const response = await this.axios.post('/auth/login', { ...credentials, mode: this.mode });
 
-		this.token = response.data.data.access_token;
+		const data = response.data.data;
+		this.token = data.access_token;
+		this.expiresAt = Date.now() + data.expires;
+
+		await this.storage.setItem('directus_access_token', this.token);
+		await this.storage.setItem('directus_access_token_expires', this.expiresAt);
 
 		if (this.mode === 'json') {
-			await this.storage.setItem('directus_refresh_token', response.data.data.refresh_token);
+			await this.storage.setItem('directus_refresh_token', data.refresh_token);
 		}
-
 		if (this.autoRefresh) {
-			setTimeout(() => this.refresh(), response.data.data.expires - 10000);
+			this.refresh();
 		}
 
 		return response.data;
 	}
 
+	/**
+	 * Refresh access token 10 seconds before expiration
+	 */
 	async refresh() {
+		clearTimeout(this.autoRefreshTimeout);
+
+		this.expiresAt = await this.storage.getItem('directus_access_token_expires');
+		if (!this.expiresAt) return;
+
+		if (Date.now() + 10000 < this.expiresAt) {
+			this.autoRefreshTimeout = setTimeout(
```

Comment:
It's possible that the user calls `refresh()` manually with `autoRefresh` disabled. In that case we shouldn't create the timeout

Reply:
added check

Reply:
Should he be able to call `refresh` if he/she is not logged in at all? `if (!this.expiresAt) return;` server as check if user is logged in

Reply:
No that's OK 👍🏻